### PR TITLE
Web Terminal 2

### PR DESCRIPTION
**Pull request by lekkas**
Connects to #27

---
This change is [Reviewable](https://reviewable.io/reviews/resin-io/resin-supervisor/111)

### CODE REVIEW

**Comment by emirotin** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[src/lib/term.coffee, line 17 \[r1\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111#-KD-XRR8TUMQ5s4YPbYd:-KD-XRR8TUMQ5s4YPbYe:428370467) ([raw file](https://github.com/resin-io/resin-supervisor/blob/8fc5808829c63cd88d1edaa1899d7eade99e2586/src/lib/term.coffee#L17)):*
Why requiring dynamically?

---

*[src/lib/term.coffee, line 26 \[r1\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111#-KD-XgCzx0a9Ix-9bcc1:-KD-XgCzx0a9Ix-9bcc2:-1978377823) ([raw file](https://github.com/resin-io/resin-supervisor/blob/8fc5808829c63cd88d1edaa1899d7eade99e2586/src/lib/term.coffee#L26)):*
can use interpolation: "shell exit, code #{code}..."

---

*[src/lib/term.coffee, line 27 \[r1\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111#-KD-XkvMnIm-hdIF8CTo:-KD-XkvMnIm-hdIF8CTp:231181382) ([raw file](https://github.com/resin-io/resin-supervisor/blob/8fc5808829c63cd88d1edaa1899d7eade99e2586/src/lib/term.coffee#L27)):*
can use object shortcut syntax: `{ signal }`

---

*[src/lib/term.coffee, line 36 \[r1\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111#-KD-Y-M6lDnJVGcVfup2:-KD-Y-M6lDnJVGcVfup3:1401322241) ([raw file](https://github.com/resin-io/resin-supervisor/blob/8fc5808829c63cd88d1edaa1899d7eade99e2586/src/lib/term.coffee#L36)):*
Does `getOrGenerateSecret` return a Bluebird promise? This method mixes promises with callbacks, can be shortened:

```
utils.getOrGenerateSecret('api')
.catch ->
  throw new Error('Invalid API key in supervisor')
.then (secret) ->
  if data.token isnt secret
    throw new Error('Authentication error')
.asCallback(cb)

```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111)*

**Comment by emirotin** [comment]
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/lib/tty.coffee, line 27 \[r1\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111#-KD-Z0NOgaS_rNBQF21H:-KD-Z0NPVHJHw8hxRNeM:317348384) ([raw file](https://github.com/resin-io/resin-supervisor/blob/8fc5808829c63cd88d1edaa1899d7eade99e2586/src/lib/tty.coffee#L27)):*
seems to be constant? I'd make it UPPERCASE, or even move to config if so

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111)*

File: docs/API.md [hunk 1]
```diff
@@ -120,11 +120,8 @@ From the app on the device:
 $ curl -X POST --header "Content-Type:application/json" \
 	--data '{"appId": <appId>}' \
 	"$RESIN_SUPERVISOR_ADDRESS/v1/spawn-tty?apikey=$RESIN_SUPERVISOR_API_KEY"
-```
-Response:
-```none
-http://124135325.ngrok.com
-```
+
+When successful, responds with an empty 200 response.
```
Comment (pcarranzav) [inline docs/API.md]: This should replace line 106

**Comment by lekkas** [comment]
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/lib/term.coffee, line 17 \[r1\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111#-KD-XRR8TUMQ5s4YPbYd:-KD-lHnOS467DSaYEn3I:1194607175) ([raw file](https://github.com/resin-io/resin-supervisor/blob/8fc5808829c63cd88d1edaa1899d7eade99e2586/src/lib/term.coffee#L17)):*
I had this code initially in tty.js and did it to save resources, then left it there. I'll move it on top with the rest require'd packages. 

---

*[src/lib/term.coffee, line 36 \[r1\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111#-KD-Y-M6lDnJVGcVfup2:-KD-yvaKpUxN6EafOqAT:-1908835108) ([raw file](https://github.com/resin-io/resin-supervisor/blob/8fc5808829c63cd88d1edaa1899d7eade99e2586/src/lib/term.coffee#L36)):*
Good point. Also adding a `return true` statement and a clarifying comment, because socketio-auth module requires cb(null, true) on success

---

*[src/lib/tty.coffee, line 27 \[r1\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111#-KD-Z0NOgaS_rNBQF21H:-KD-pvi5HT51kt9Mn2AM:-1418795847) ([raw file](https://github.com/resin-io/resin-supervisor/blob/8fc5808829c63cd88d1edaa1899d7eade99e2586/src/lib/tty.coffee#L27)):*
Right, moving it to config

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111)*

**Comment by lekkas** [comment]
Review status: 0 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[docs/API.md, line 124 \[r1\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111#-KD-4SDlcYlxzhM_g7sT-r1-124:-KD0-zS6Xa6dCeOi6EKq:1036516611) ([raw file](https://github.com/resin-io/resin-supervisor/blob/8fc5808829c63cd88d1edaa1899d7eade99e2586/docs/API.md#L124)):*
Done, also added (Empty response) to be consistent with other parts of the API doc

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111)*

**Comment by emirotin** [comment]
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/lib/tty.coffee, line 42 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111#-KD01gIsJXRVSoZ2wl08:-KD01gIt4nfkmxhq8p1x:572201699) ([raw file](https://github.com/resin-io/resin-supervisor/blob/7ab83ef44ba2a656380bb283473a2b2fb4e48641/src/lib/tty.coffee#L42)):*
Looks like this uses the old syntax for `promisify` that accepts context. It's deprecated in the newest version of Bluebird (now it accepts it as `{ context: ... }`), I suggest using `server.listen.bind(server)` — I think this should work in every version

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111)*

**Comment by lekkas** [comment]
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/lib/tty.coffee, line 42 \[r3\]](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111#-KD01gIsJXRVSoZ2wl08:-KD4iipWMouq8--0Typ0:1949392400) ([raw file](https://github.com/resin-io/resin-supervisor/blob/7ab83ef44ba2a656380bb283473a2b2fb4e48641/src/lib/tty.coffee#L42)):*
Good catch Eugene, done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/resin-io/resin-supervisor/111)*

**Comment by pcarranzav** [comment]
@lekkas I guess we should close this now the terminal was implemented outside the supervisor?

**Comment by lekkas** [comment]
Yeah, I've also saved the branch locally just in case. It was a good exercise nevertheless! :)

**Pull request closed without merging**